diesel/src/query_source/joins.rs

```diff
@@ -21,22 +31,30 @@ impl<Left, Right, Kind> Join<Left, Right, Kind> {
             kind: kind,
         }
     }
+
+    #[doc(hidden)]
+    pub fn on<On>(self, on: On) -> JoinOn<Self, On> {
+        JoinOn {
+            join: self,
+            on: on,
```

Comment:
Just wondering: On has no constraints, but it will probably always need to be `On: Expression`. Could adding this here (or in `fn join`) give better/earlier errors?